src/backend/distributed/metadata/node_metadata.c

```diff
@@ -1131,143 +839,246 @@ PrimaryNodeForGroup(int32 groupId, bool *groupContainsNodes)
-		WorkerNode *workerNode =
-			FindWorkerNodeAnyCluster(node->workerName, node->workerPort);
-		if (workerNode == NULL)
-		{
-			ereport(ERROR, (errmsg("node at \"%s:%u\" does not exist", node->workerName,
-								   node->workerPort)));
-		}
+		StringInfo metadatasyncCommand = makeStringInfo();
+		appendStringInfo(metadatasyncCommand, CITUS_MARK_NODE_NOT_SYNCED,
+						 parentSessionPid, workerNode->nodeId);
+		commandList = lappend(commandList, metadatasyncCommand->data);
+	}
 
-		/* both nodes should be the same */
-		Assert(workerNode->nodeId == node->nodeId);
+	SendCommandListToWorkerOutsideTransactionWithConnection(connection, commandList);
+	CloseConnection(connection);
+}
 
-		/*
-		 * Delete existing reference and replicated table placements on the
-		 * given groupId if the group has been disabled earlier (e.g., isActive
-		 * set to false).
-		 *
-		 * Sync the metadata changes to all existing metadata nodes irrespective
-		 * of the current nodes' metadata sync state. We expect all nodes up
-		 * and running when another node is activated.
-		 */
-		if (!workerNode->isActive && NodeIsPrimary(workerNode))
+
+/*
+ * SetNodeMetadata sets isactive, metadatasynced and hasmetadata flags locally
+ * and, if required, remotely.
+ */
+static void
+SetNodeMetadata(MetadataSyncContext *context, bool localOnly)
+{
+	/* do not execute local transaction if we collect commands */
+	if (!MetadataSyncCollectsCommands(context))
+	{
+		List *updatedActivatedNodeList = NIL;
+
+		WorkerNode *node = NULL;
+		foreach_ptr(node, context->activatedWorkerNodeList)
 		{
-			bool localOnly = false;
-			DeleteAllReplicatedTablePlacementsFromNodeGroup(workerNode->groupId,
-															localOnly);
+			node = SetWorkerColumnLocalOnly(node, Anum_pg_dist_node_isactive,
+											BoolGetDatum(true));
+			node = SetWorkerColumnLocalOnly(node, Anum_pg_dist_node_metadatasynced,
+											BoolGetDatum(true));
+			node = SetWorkerColumnLocalOnly(node, Anum_pg_dist_node_hasmetadata,
+											BoolGetDatum(true));
+
+			updatedActivatedNodeList = lappend(updatedActivatedNodeList, node);
 		}
 
-		workerNode =
-			SetWorkerColumnLocalOnly(workerNode, Anum_pg_dist_node_isactive,
-									 BoolGetDatum(true));
+		/* reset activated nodes inside metadataSyncContext afer local update */
+		SetMetadataSyncNodesFromNodeList(context, updatedActivatedNodeList);
+	}
 
-		/* TODO: Once all tests will be enabled for MX, we can remove sync by default check */
-		bool syncMetadata = EnableMetadataSync && NodeIsPrimary(workerNode);
-		if (syncMetadata)
+	if (!localOnly && EnableMetadataSync)
+	{
+		WorkerNode *node = NULL;
+		foreach_ptr(node, context->activatedWorkerNodeList)
 		{
-			/*
-			 * We are going to sync the metadata anyway in this transaction, so do
-			 * not fail just because the current metadata is not synced.
-			 */
-			SetWorkerColumn(workerNode, Anum_pg_dist_node_metadatasynced,
-							BoolGetDatum(true));
+			SetWorkerColumnViaMetadataContext(context, node,
```

Comment:
In non-tx mode, `SetWorkerColumnViaMetadataContext` forces new connections. And, say if you have 10 nodes, we establish 10x10x3 connections. 

Can we at least merge the 3 commands into a single one?